src/multiset.rs

```diff
@@ -366,17 +356,11 @@ where
     /// assert_eq!(1, combined.count_of(&4));
     /// assert_eq!(0, combined.count_of(&5));
     /// ```
-    fn add(self, rhs: HashMultiSet<T>) -> HashMultiSet<T> {
-        let mut ret: HashMultiSet<T> = HashMultiSet::new();
-        for val in self.distinct_elements() {
-            let count = self.count_of(val);
-            ret.insert_times((*val).clone(), count);
+    fn add(mut self, rhs: HashMultiSet<T>) -> HashMultiSet<T> {
```

Comment:
Taking `mut self` and returning the modified self is closer to the intent of `AddAssign` and `SubAssign`.

What do you think about keeping `Add` and `Sub` immutable, and optionally add new *Assign implementations?

Reply:
Because the `Add` and `Sub` traits take `self` by value (not reference), you wouldn't be able to access the old value of `self` anyways...so I don't think there's much of a difference.

The `mut` in the signature doesn't actually change the input type, it's equivalent to the following:

```rust
fn add(self, rhs: HashMultiSet<T>) -> HashMultiSet<T> {
    let mut ret = self;
    /// ...
}
```

~Honestly, I'm not completely sure why `AddAssign` isn't a marker trait that automatically desugars `x += y` into `x = x + y` but doesn't require a new method...~

I guess the reason is because `Add` requires that you consume `self`, while `AddAssign` requires that you take self by mutable reference, which is different.

Since having an `Add` implementation that *doesn't* take self by value could also be useful, we could always implement `Add` for `&HashMultiSet<T>`.

**NOTE**: Everything I said about `Add*` applies to `Sub*`.

Reply:
The implementation proposed by this PR is way better for performance, and simpler.

Reply:
Yes, since `self` and `mut self` both consume. There's no difference there.

Reply:
Ideally since rhs is also consumed, one could check which of the hashmaps is larger in order to merge the smallest one in the largest one, minimizing the number of operations.